Public/Src/Engine/Processes/UnSandboxedProcess.cs

```diff
@@ -210,37 +186,21 @@ public virtual async Task<SandboxedProcessResult> GetResultAsync()
         {
             Contract.Requires(Started);
 
-            // 1: wait until process exits or process timeout is reached
-            LogProcessState("Waiting for process to exit");
-            var finishedTask = await Task.WhenAny(Task.Delay(ProcessInfo.Timeout.Value), WhenExited);
-            m_exitTime = DateTime.UtcNow;
-
-            // 2: kill process if timed out
-            var timedOut = finishedTask != WhenExited;
-            if (timedOut)
-            {
-                LogProcessState($"Process timed out after {ProcessInfo.Timeout.Value}; it will be forcefully killed.");
-                await KillAsync();
-            }
+            await m_processExecutor.WaitForExitAsync();

```

Comment:
Although I like it better how it looks now, I am not 100% sure that the new semantics of this method is going to always be correct.

Before, we executed steps 1, 2, 3, 4 in order; now the order changed to 1, 2, 4, 3 (i.e., wait for stdout/stderr before waiting for process reports).  In most cases, this order doesn't matter.  But, consider the following scenario:
  - process `A` starts
  - process `A` spawns a child process `a` which is declared as "allowed to survive nested process"
  - `A` exits; since `a` is still running (and will never finish), stdout/stderr of `A` may still be opened (because child processes often inherit them)
It is `GetReportsAsync()` that checks if any child processes are still running and, in particular, handles allowed surviving child processes.  With this change, that code may never kick in. 

Adding @Kr1sso.  It would probably make sense to get some feedback from him on this too.

Reply:
I'll pass an argument to `WaitForExitAsync` to get the reports in the middle.

---
In reply to: [277343477](https://github.com/Microsoft/BuildXL/pull/159#discussion_r277343477) [](ancestors = 277343477)

Reply:
👍 Thanks.